guacamole/src/main/webapp/app/navigation/services/userPageService.js

```diff
@@ -140,13 +144,20 @@ angular.module('navigation').factory('userPageService', ['$injector',
         var deferred = $q.defer();
 
         // Resolve promise using home page derived from root connection groups
-        dataSourceService.apply(
+        var getRootGroups = dataSourceService.apply(
             connectionGroupService.getConnectionGroupTree,
             authenticationService.getAvailableDataSources(),
             ConnectionGroup.ROOT_IDENTIFIER
-        )
-        .then(function rootConnectionGroupsRetrieved(rootGroups) {
-            deferred.resolve(generateHomePage(rootGroups));
+        );
+        var getPermissionSets = dataSourceService.apply(
+            permissionService.getPermissions,
+            authenticationService.getAvailableDataSources(),
+            authenticationService.getCurrentUsername()
+        );
+
+        $q.all([getRootGroups,getPermissionSets])
+        .then(function rootConnectionGroupsPermissionsRetrieved(data) {
+            deferred.resolve(generateHomePage(data[0],data[1]));
```

Comment:
According to [the AngularJS documentation for `$q.all()`](https://code.angularjs.org/1.3.16/docs/api/ng/service/$q#all), the function can accept either an array (in which case `data` will be a parallel array, as you use here) or an object (in which case `data` will be an object with identical keys pointing to the results). For the sake of readability, I suggest switching to the latter.

Reply:
Objectified.